File: leetcode/leetcode_1.py
from typing import List
class Solution: #Brute force
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        for index, value in enumerate(nums):
            for inner_index, inner_value in enumerate(nums[index + 1:]):
                if nums[index] + inner_value == target:
                    return [index, inner_index + index + 1]

    def twoSum(self, nums: List[int], target: int) -> List[int]:# using in
        for i, n in enumerate(nums):
            complement = target - n

            if complement in nums[i + 1:]:
                return [nums.index(n), nums[i+1:].index(complement) + (i + 1)]

    def twoSum(self, nums: List[int], target: int) -> List[int]:  # dict
        nums_map = {}
        # switch index and value
        for i ,num in enumerate(nums):
            nums_map[num] = i

        for i, num in enumerate(nums):
            if target - num in nums_map and i != nums_map[target - num]:
                return [i, nums_map[target - num]]

File: leetcode/leetcode_1.java
import java.util.HashMap;
import java.util.ArrayList;
class Solution {
    public int[] twoSum(int[] nums, int target) {
        int[] r = new int[] {0,0}; 
        HashMap <Integer, ArrayList<Integer>> h = new HashMap<>();
        for (int i =0; i < nums.length; i ++){
            ArrayList<Integer> arr = h.containsKey(nums[i]) ? h.get(nums[i]) : new ArrayList<Integer> ();
            arr.add(i);
            h.put(nums[i], arr);
        }

        for(int num : nums){
            if (h.containsKey(target - num)){
                if (target-num == num){
                   if(h.get(num).size() >= 2){
                    r[0]  = h.get(num).get(0);
                    r[1] = h.get(num).get(1);
                    return r;
                    }
                }
                else{
                    r[0] = h.get(num).get(0);
                    r[1] = h.get(target-num).get(0);
                    return r;
                }
            }
        }
        return r;
    }
}
